Feature/141-aai-compose-and-config (#162)
puzzle/unilu-pruefungsabfrage: shibboleth/README.md

![SwitchAAI](./doc/SwitchAAI.drawio.png)

## Create & Run Docker Image Locally
Preparation:

Provided certificates can be downloaded [here](https://files.puzzle.ch/apps/files/files/7229088?dir=/Styles%20Uni%20Luzern/Certificates). For instructions where to place files or how to generate them read
the [Generate New Certificates](#generate-new-certificates) section.

1. Copy the provided httpd certificate and key to the test or prod dir or generate them. For local development it's easiest to
use self signed certificates.

> 2. Copy the provided sp certificate and key to the test or prod dir. It is also possible to generate them but we currently

Reviewer: ```suggestion
2. Copy the provided sp certificate and key to the  `/shibboleth/secrets/test`  or `/shibboleth/secrets/prod` directory. It is also possible to generate them but we currently
```

need to use the provided ones.

> 3. Download the `SWITCHaaiRootCA.crt.pem` and place it in the `/secrets` dir (not in the prod or test dir!). This file cannot

Reviewer: ```suggestion
3. Download the `SWITCHaaiRootCA.crt.pem` and place it into the `/secrets` dir (not in the `prod`  or `test` dir!). This file cannot
```

be generated.
4. Generate the keystore for the httpd certificate and key by going into the test or prod dir and then using the correct command:
   - Prod: `openssl pkcs12 -export -in httpd.crt.pem -inkey httpd.key.pem -name unilu-eft -out httpd.keystore.p12`. Choose a good password
   and remember it, as you will need it to run the application.
   - Test: `openssl pkcs12 -export -in httpd.crt.pem -inkey httpd.key.pem -name unilu-eft-test -out httpd.test.keystore.p12`. Here you need to use the
   password in the application-test.yml.
5. Add aliases for `edview.unilu.ch` and `edview-test.unilu.ch` to your [host file](https://linuxhandbook.com/etc-hosts-file/).

There are two docker compose files to run the application. The `docker-compose.yml` file is the dev config which uses the secrets

> in `/shibboleth/secrets/test`. This config rebuilds the service provider image every time and the application is started with a maven command within the container.

Reviewer: ```suggestion
in `/shibboleth/secrets/test`. This config rebuilds the service provider image every time and the application is started with a Maven command within the container.
```
Typo

the [Generate New Certificates](#generate-new-certificates) section.

1. Copy the provided httpd certificate and key to the test or prod dir or generate them. For local development it's easiest to
use self signed certificates.
2. Copy the provided sp certificate and key to the test or prod dir. It is also possible to generate them but we currently
need to use the provided ones.
3. Download the `SWITCHaaiRootCA.crt.pem` and place it in the `/secrets` dir (not in the prod or test dir!). This file cannot
be generated.
4. Generate the keystore for the httpd certificate and key by going into the test or prod dir and then using the correct command:
   - Prod: `openssl pkcs12 -export -in httpd.crt.pem -inkey httpd.key.pem -name unilu-eft -out httpd.keystore.p12`. Choose a good password
   and remember it, as you will need it to run the application.
   - Test: `openssl pkcs12 -export -in httpd.crt.pem -inkey httpd.key.pem -name unilu-eft-test -out httpd.test.keystore.p12`. Here you need to use the
   password in the application-test.yml.

>    Attention: If you are running rootless docker you have to adjust the file permission of the keystore so others can read it. Else the file

Reviewer: ```suggestion
   **Attention**: If you are running rootless docker you have to adjust the file permission of the keystore so others can read it. Else the file
```

|-------------------|-----------------------------------------------------------------------------------------------------------|-------------------------|-----------|
| TAG               | The tag of the application and service provider image that should be used                                 | -                       | Yes       |
| RESOURCE_LOCATION | Path to the directory containing the subject directories with the exams in them                           | -                       | Yes       |
| KS_PW             | The password which was while generating the keystore                                                      | -                       | Yes       |
| KS_TYPE           | Tells the application what type of keystore is used                                                       | PKCS12                  | No        |
| KS_NAME           | The full name including file extension of the keystore                                                    | httpd.keystore.p12      | No        |
| KEY_ALIAS         | Sets the alias that identifies the key in the keystore                                                    | unilu-eft               | No        |

> | SEC_LOCATION      | The location where certificates, keys and keystore are stored. Value relative to prod docker compose file | shibboleth/secrets/prod | No        |

Reviewer: ```suggestion
| SEC_LOCATION      | The location where certificates, keys and keystore are stored. Is relative to the  docker compose file | shibboleth/secrets/prod | No        |
```

To run this config execute the following command:

`RESOURCE_LOCATION=<resource-path> KS_PW=<keystore-password> TAG=<image-tag> docker compose -f docker-compose-prod.yml up --build --force-recreate`

When you start a session within the docker image you can find the metadata file from `AAI Test` or `SWITCH AAI` under
`/var/cache/shibboleth/` which will be refreshed every hour.

## Generate New Certificates

| Environment | Default secret location  |
|-------------|--------------------------|
| Prod        | /shibboleth/secrets/prod |
| Test        | /shibboleth/secrets/test |

The following steps can be done for `test` and / or `prod` environment. The commands allow the parameter `test` or
`prod`, which will generate the certificates in `/shibboleth/secrets/test` or `/shibboleth/secrets/prod` accordingly.
The default environment is `test`.

> 1. Create the host certificate for Apache HTTP/S server using the command `httpd-certificate`. The configuration is

Reviewer: ```suggestion
1. Create the host certificate for Apache HTTP/S server using the command `./httpd-certificate`. The configuration is
```

   defined in `httpd.crt.conf`. The result is the public certificate `httpd.crt` and its private key file
   `httpd.key.pem`.

> 2. Create the Shibboleth Service Provider certificate using the command `sp-certificate`. The configuration is defined

Reviewer: ```suggestion
2. Create the Shibboleth Service Provider certificate using the command `./sp-certificate`. The configuration is defined
```
